sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/sns/SnsIO.java

```diff
@@ -0,0 +1,299 @@
+    public Write withRetryDelay(Duration retryDelay) {
+      return builder().setRetryDelay(retryDelay).build();
+    }
+
+    /** Tuple tag to store results. Mandatory field. */
+    public Write withResultOutputTag(TupleTag<PublishResult> results) {
+      return builder().setResultOutputTag(results).build();
+    }
+
+    @Override
+    public PCollectionTuple expand(PCollection<PublishRequest> input) {
+      checkArgument(getTopicName() != null, "withTopicName() is required");
+      final PCollectionTuple tuple =
+          input.apply(
+              ParDo.of(new SnsWriterFn(this))
+                  .withOutputTags(getResultOutputTag(), TupleTagList.empty()));
+      return tuple;
+    }
+
+    private static class SnsWriterFn extends DoFn<PublishRequest, PublishResult> {
+
+      private static final Logger LOG = LoggerFactory.getLogger(SnsWriterFn.class);
+      private static final Counter SNS_WRITE_FAILURES =
+          Metrics.counter(SnsWriterFn.class, "SNS_Write_Failures");
+
+      private final SnsIO.Write spec;
+      private transient AmazonSNS producer;
+      private transient RetryPolicy retryPolicy;
+
+      public SnsWriterFn(SnsIO.Write spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void setup() throws Exception {
+
+        //Initialize SnsPublisher
+        producer = spec.getAWSClientsProvider().createSnsPublisher();
+        retryPolicy =
+            new RetryPolicy()
+                .retryOn(ServiceUnavailableException.class)
+                .withBackoff(
+                    spec.getRetryDelay().toMillis(),
+                    spec.getMaxDelay().toMillis(),
+                    TimeUnit.MILLISECONDS)
+                .withMaxRetries(spec.getMaxRetries());
+        checkArgument(
+            topicExists(producer, spec.getTopicName()),
+            "Topic %s does not exist",
+            spec.getTopicName());
+      }
+
+      private PublishResult getResult(PublishRequest request) throws Exception {
+        PublishResult result = producer.publish(request);
+        int statusCode = result.getSdkHttpMetadata().getHttpStatusCode();
+        if (statusCode == 200) {
+          return result;
+        } else if (statusCode == 503) {
+          throw new ServiceUnavailableException(
+              "Service Temporarily unavailable. Will retry publishing " + request.getMessage(),
+              new RuntimeException("Got " + statusCode));
+        } else {
+          throw new SnsWriteException(
+              "Unable to publish message " + request.getMessage(),
+              new RuntimeException("Got " + statusCode));
+        }
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) {
+        PublishRequest request = context.element();
+        Failsafe.with(retryPolicy)
+            .run(
+                () -> {
+                  try {
+                    PublishResult pr = getResult(request);
+                    LOG.info(
```

Comment:
make LOG.debug or remove (considering the payload may contain sensitive data) for more details on why you can check the [PTransform guide](https://beam.apache.org/contribute/ptransform-style-guide/#logging)